src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp

```diff
@@ -8117,6 +8117,17 @@ CTranslatorExprToDXL::GetInputSegIdsArray(CExpression *pexprMotion)
 	if (CUtils::FDuplicateHazardMotion(pexprMotion) ||
 		CDistributionSpec::EdtTaintedReplicated == pds->Edt())
 	{
+		// FIXME: If there is any 'unpaired' CTE Consumer beneath, fallback to Postgres optimizer.
+		// Related to: https://github.com/greenplum-db/gpdb/issues/13039
+
+		if (!CUtils::CanReadFromOneSegment(m_mp, pexprMotion))
+		{
+			GPOS_RAISE(
+				gpdxl::ExmaDXL, gpdxl::ExmiExpr2DXLUnsupportedFeature,
+				GPOS_WSZ_LIT(
+					"CTE Consumer without the appropriate CTE Producer under a duplicate-hazard motion or a tainted replicated node"));
+		}
+
```

Comment:
It's better to move your check block under the comment below because of that comment describes the case of a whole branch under condition `CUtils::FDuplicateHazardMotion(pexprMotion) || CDistributionSpec::EdtTaintedReplicated == pds->Edt()` and your block falls under this check.

We might conclude that after locality validation routine `CUtils::ValidateCTEProducerConsumerLocality()` we end up to the plan with strong correspondence between CTE producers and CTE consumers. But in current branch we in fact transform the Motion subplan from `EeltSegments` locality value to `EeltSingleton` one and thereby violate the check result from previous `CUtils::ValidateCTEProducerConsumerLocality()` call if CTE consumers in current subplan have corresponding producers somewhere externally. Therefore, it's enough to check on existence of external CTE consumers in current Motion subplan to verify that we don't break the `CUtils::ValidateCTEProducerConsumerLocality()` check.

The reasoning above might be posted as comment to prove correctness of your check.

Reply:
Renamed the function and extended PR description, commit message and in-code comments. Thanks!